### DISCUSSION TITLE
请教 OSR 模式下的性能问题

### DISCUSSION THREAD

**Discussion by catcheroftime** [CefView/QCefView#147] [General]
因为我的开发需要，我需要使用 QCefView 的 OSR 模式，但是在实际使用过程中发现，
在 4K 场景下使用会鼠标移动网页中的元素，动画等会出现很明显的卡顿，分析问题的时候发现
cef 本身的 OnPaint() 函数回调函数的速度一般都在 80 ms 左右，此外拿到图片后在 QCefView 中渲染一张 4K 的图片也需要 40+ ms
这一块有优化的空间吗？希望可以提供一些思路

我做了一些如下的尝试
1. 更新到最新的 cef 版本 105.3.36， 新版本并没有明显的优化
2. 设置 disable-gpu，disable-gpu-compositing  查看官方的 issues 里建议是禁用GPU并使用软件渲染来获得最佳帧速率/性能权衡，效果也不是很好
   https://bitbucket.org/chromiumembedded/cef/issues/2046/improve-gpu-readback-performance-for-osr
   https://bitbucket.org/chromiumembedded/cef/pull-requests/96
3. 因为在非全屏，较小界面(1920x1080)的情况下，动画之类性能表现还是可以达到要求的，所以尝试 CCefClientDelegate 将 getViewRect() 获取的尺寸调整到一个比较小的数值比如 1920x1080， 这样 OnPaint() 返回的图片的长宽也是 1920x1080，渲染的时候将图片根据实际情况放缩，但是这里需要将鼠标的坐标重新做个映射关系，但是 4K全屏下，虽然 OnPaint() 返回的图片大小确实是 1920x1080，但是调用间隔还是 80ms 左右，似乎没有生效，离屏渲染后台解析图片的逻辑和想象中不太一样
4. 此外，因 QCefView 中接收新的图片数据的时候需要等渲染结束，一开时候怀疑是渲染锁的问题，尝试过用队列缓存 OnPaint() 回调的图片，在队列不为空的时候通知界面自己取数据绘制，但是问题实际还是 OnPaint() 函数回调函数的速度上

**Comment by tishion** [comment]
你不用尝试了，4K下面无法达到你的要求，因为现在OSR模式下跟Qt的结合方式是实用的Qt的Software Rasterizer作为渲染引擎，CEF内部虽然是硬件加速生成每一帧图像，但是要把CEF生成的图像绘制到QT窗口这个过程目前没有做加速，所以高分辨率下，性能肯定很差。建议使用NCW模式吧
